### PR TITLE
Test runner improvements

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1291) |

#### Main changes
- Re-add camera tests.
- Add thread stop conditions on exit.
- Use `pytest-xdist` to run tests in multiple CPUs and speed up test execution.
- Use `pytest.mark.parameterized` instead of `parameterized` module.
- Convert `unittest.TestCases` into `pytest` tests.
- Cache python dependencies to reduce workflow run times.
- Cache shape predictors used in face detector tests.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: pitop/robotics/navigation/core/measurement_scheduler.py [hunk 1]
```diff
@@ -43,9 +49,10 @@ def loop(self, s, previous_time):
...
-            current_time + self._measurement_dt,
-            1,
-            self.loop,
-            (s, current_time),
-        )
+        if self._continue_processing:
+            s.enterabs(
+                current_time + self._measurement_dt,
+                1,
+                self.loop,
```
Comment (angusjfw) [inline pitop/robotics/navigation/core/measurement_scheduler.py:56]: nice work getting all these threads to have proper exit conditions. unrelated but this use of recursion in this scheduler has caught my eye as something that could cause a stack overflow? maybe something to look into in future.

File: pitop/robotics/navigation/navigation_controller.py [hunk 2]
```diff
@@ -129,6 +128,10 @@ def go_to(
 
         return self
 
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if self._nav_thread.is_alive():
```
Comment (angusjfw) [inline pitop/robotics/navigation/navigation_controller.py:132]: should you set `self.navigator._stop_triggered = True` or something here to ensure this thread will actually join and allow the process to exit?

File: .github/workflows/test-all-commits.yml [hunk 3]
```diff
@@ -21,13 +21,53 @@ jobs:
...
+           ${{ runner.os }}-sdk-shape-predictors-
+
      - name: Install pip dependencies
        run: pip3 install -r tests/requirements.txt
 
      - name: Run Python tests and generate coverage report
        run: |
-         pytest --verbose --cov-report term-missing --cov=pitop
-         coverage xml
+         python3 -m pytest --verbose --cov-report term-missing --cov=pitop --numprocesses auto
```
Comment (angusjfw) [inline .github/workflows/test-all-commits.yml:69]: can we set this numprocesses thing in a pytest config file so that it is used for local test runs also?

File: tests/test_camera.py [hunk 4]
```diff
@@ -91,8 +90,18 @@ def test_current_frame_opencv(self):
...
+        timeout = 0
+        MAX_TIMEOUT = 5
+        sleep_time = 0.1
+        while frame is None and timeout < MAX_TIMEOUT:
+            sleep(sleep_time)
+            timeout += sleep_time
+            frame = c.current_frame()
+
+        if timeout >= MAX_TIMEOUT:
+            raise TimeoutError
```
[the remark this answers is not in the document]
Comment (jcapona) [reply tests/test_camera.py]: ended up using a `wait_until` helper we had, good catch

**Review by angusjfw — APPROVED** [review]

File: tests/test_lock.py [hunk 5]
```diff
@@ -38,6 +40,7 @@ def test_chmod_is_called_if_file_doesnt_exist(self, exists_mock):
         exists_mock.assert_called_once_with(self.lock_file_path)
         self.chmod_mock.assert_called_once_with(self.lock_file_path, 146)
 
+    @pytest.mark.flaky(reruns=3)
```
[the remark this answers is not in the document]
Comment (m-roberts) [reply tests/test_lock.py:43]: Wow, nice
